robot_state_publisher failed to terminate '5' seconds after receiving 'SIGINT', escalating to 'SIGTERM'

I am trying to run the webots_ros2_universal_robots package though running into, what seems to be, a robot_state_publisher error. I am using Ubuntu 20.04 LTS running ROS2 Foxy Desktop. I installed webots from Debian Packages with APT and installed the webots_ros2 packages from source [following these instructions](http://wiki.ros.org/webots_ros2) . 

When running the following example [ros2 launch webots_ros2_universal_robot universal_robot.launch.py] I get the error which I posted below. I went though the files and made the python scripts executables [ chmod +x filename.py], rebuilt my workspace, though the same error occurs. 

Prior to this, I installed webots Debian Packages with APT and webots_ros2 packages also with APT. Everything was working fine, and I was able to run the webot_ros2_universal_robot demos. I needed to modify the webot_ros2_universal_robot packages, so I created a workspace and installed them from source which is when I first ran into the error. I uninstalled the previous webot_ros2 packages installed with APT, thinking there were conflicting file paths, though that did not fix the error. I even created a new Virtual environment and installed webots debian packages and webots_ros2 from source. However the same error is present. 

Any help would be greatly appreciated. 

**ERROR MESSAGE** 

iego@diego-virtual-machine:~$ source /opt/ros/foxy/setup.bash 
diego@diego-virtual-machine:~$ cd ros2_ws/
diego@diego-virtual-machine:~/ros2_ws$ source install/setup.bash 
diego@diego-virtual-machine:~/ros2_ws$ ros2 launch webots_ros2_universal_robot universal_robot.launch.py
[INFO] [launch]: All log files can be found below /home/diego/.ros/log/2020-09-21-12-15-48-999896-diego-virtual-machine-2264
[INFO] [launch]: Default logging verbosity is set to INFO
Task exception was never retrieved
future: <Task finished name='Task-2' coro=<LaunchService._process_one_event() done, defined at /opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py:271> exception=TypeError("Can't mix strings and bytes in path components")>
Traceback (most recent call last):
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 273, in _process_one_event
    await self.__process_event(next_event)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 293, in __process_event
    visit_all_entities_and_collect_futures(entity, self.__context))
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  [Previous line repeated 3 more times]
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 38, in visit_all_entities_and_collect_futures
    sub_entities = entity.visit(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/action.py", line 108, in visit
    return self.execute(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/execute_process.py", line 775, in execute
    self.__expand_substitutions(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/execute_process.py", line 635, in __expand_substitutions
    cmd = [perform_substitutions(context, x) for x in self.__cmd]
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/execute_process.py", line 635, in <listcomp>
    cmd = [perform_substitutions(context, x) for x in self.__cmd]
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/perform_substitutions_impl.py", line 26, in perform_substitutions
    return ''.join([context.perform_substitution(sub) for sub in subs])
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/perform_substitutions_impl.py", line 26, in <listcomp>
    return ''.join([context.perform_substitution(sub) for sub in subs])
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_context.py", line 184, in perform_substitution
    return substitution.perform(self)
  File "/home/diego/ros2_ws/install/webots_ros2_core/lib/python3.8/site-packages/webots_ros2_core/webots_launcher.py", line 98, in perform
    webots_path = get_webots_home()
  File "/home/diego/ros2_ws/install/webots_ros2_core/lib/python3.8/site-packages/webots_ros2_core/utils.py", line 59, in get_webots_home
    if os.path.isdir(path) and get_webots_version(path) == version:
  File "/home/diego/ros2_ws/install/webots_ros2_core/lib/python3.8/site-packages/webots_ros2_core/utils.py", line 140, in get_webots_version
    versionFile = os.path.join(path if path is not None else get_webots_home(),
  File "/usr/lib/python3.8/posixpath.py", line 90, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/lib/python3.8/genericpath.py", line 155, in _check_arg_types
    raise TypeError("Can't mix strings and bytes in path components") from None
TypeError: Can't mix strings and bytes in path components
Task exception was never retrieved
future: <Task finished name='Task-8' coro=<LaunchService._process_one_event() done, defined at /opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py:271> exception=RuntimeError('Signal event received before subprocess transport available.')>
Traceback (most recent call last):
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 273, in _process_one_event
    await self.__process_event(next_event)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/launch_service.py", line 293, in __process_event
    visit_all_entities_and_collect_futures(entity, self.__context))
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 38, in visit_all_entities_and_collect_futures
    sub_entities = entity.visit(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/action.py", line 108, in visit
    return self.execute(context)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/opaque_function.py", line 75, in execute
    return self.__function(context, *self.__args, **self.__kwargs)
  File "/opt/ros/foxy/lib/python3.8/site-packages/launch/actions/execute_process.py", line 421, in __on_signal_process_event
    raise RuntimeError('Signal event received before subprocess transport available.')
RuntimeError: Signal event received before subprocess transport available.
[INFO] [robot_state_publisher-1]: process started with pid [2286]
**[robot_state_publisher-1] Parsing robot urdf xml string.
[robot_state_publisher-1] [INFO] [1600715749.248895892] [robot_state_publisher]: got segment 
[ERROR] [robot_state_publisher-1]: process[robot_state_publisher-1] failed to terminate '5' seconds after receiving 'SIGINT', escalating to 'SIGTERM'
[INFO] [robot_state_publisher-1]: sending signal 'SIGTERM' to process[robot_state_publisher-1]
[ERROR] [robot_state_publisher-1]: process has died [pid 2286, exit code -15, cmd '/opt/ros/foxy/lib/robot_state_publisher/robot_state_publisher --ros-args --params-file /tmp/launch_params_dcugksof'].**

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -50,7 +50,7 @@ def get_webots_home(version=get_required_webots_version()):
     # search first with the environment variables
     environVariables = ['ROS2_WEBOTS_HOME', 'WEBOTS_HOME']
     for variable in environVariables:
-        if variable in os.environ and os.path.isdir(variable) and get_webots_version(os.environ[variable]) == version:
+        if variable in os.environ and os.path.isdir(os.environ[variable]) and get_webots_version(os.environ[variable]) == version:
             os.environ['WEBOTS_HOME'] = os.environ[variable]
             return os.environ[variable]
     # then using the 'which' command

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -139,7 +139,7 @@ def get_webots_version_major_number():
 
 def get_webots_version(path=None):
     """Webots version as a string."""
-    versionFile = os.path.join(path if path is not None else get_webots_home(), 
+    versionFile = os.path.join(path if path is not None else get_webots_home(),
                                'resources', 'version.txt')
     if not os.path.isfile(versionFile):
         return None

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -45,9 +45,11 @@ def get_required_webots_version_short():
     return make_short_version(get_required_webots_version())
 
 
-def get_webots_home(version=get_required_webots_version()):
+def get_webots_home(version=None):
     """Path to the Webots installation directory."""
     # search first with the environment variables
+    if version is None:
+        version = get_required_webots_version()
     environVariables = ['ROS2_WEBOTS_HOME', 'WEBOTS_HOME']
     for variable in environVariables:
         if variable in os.environ and os.path.isdir(variable) and get_webots_version(os.environ[variable]) == version:
@@ -56,6 +58,7 @@ def get_webots_home(version=get_required_webots_version()):
     # then using the 'which' command
     try:
         path = os.path.split(os.path.abspath(subprocess.check_output(['which', 'webots'])))[0]
+        path = path.decode('utf-8')
         if os.path.isdir(path) and get_webots_version(path) == version:
             os.environ['WEBOTS_HOME'] = path
             return path
@@ -126,7 +129,8 @@ def append_webots_python_lib_to_path():
 
 def get_webots_version_major_number():
     """Webots major version as an integer."""
-    versionString = get_webots_version()
+    webots_home = get_webots_home()
+    versionString = get_webots_version(webots_home)
     if versionString is None:
         return 0
     match = re.match(r'R(\d*).*', versionString)
@@ -135,10 +139,9 @@ def get_webots_version_major_number():
     return 0
 
 
-def get_webots_version(path=None):
+def get_webots_version(path):
     """Webots version as a string."""
-    versionFile = os.path.join(path if path is not None else get_webots_home(),
-                               'resources', 'version.txt')
+    versionFile = os.path.join(path, 'resources', 'version.txt')
     if not os.path.isfile(versionFile):
         return None
     with open(versionFile, 'r') as f:

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -50,7 +50,8 @@ def get_webots_home(version=get_required_webots_version()):
     # search first with the environment variables
     environVariables = ['ROS2_WEBOTS_HOME', 'WEBOTS_HOME']
     for variable in environVariables:
-        if variable in os.environ and os.path.isdir(os.environ[variable]) and get_webots_version(os.environ[variable]) == version:
+        if variable in os.environ and os.path.isdir(os.environ[variable]) and \
+            get_webots_version(os.environ[variable]) == version:
             os.environ['WEBOTS_HOME'] = os.environ[variable]
             return os.environ[variable]
     # then using the 'which' command
@@ -127,8 +128,7 @@ def append_webots_python_lib_to_path():
 
 def get_webots_version_major_number():
     """Webots major version as an integer."""
-    webots_home = get_webots_home()
-    versionString = get_webots_version(webots_home)
+    versionString = get_webots_version()
     if versionString is None:
         return 0
     match = re.match(r'R(\d*).*', versionString)
@@ -137,9 +137,10 @@ def get_webots_version_major_number():
     return 0
 
 
-def get_webots_version(path):
+def get_webots_version(path=None):
     """Webots version as a string."""
-    versionFile = os.path.join(path, 'resources', 'version.txt')
+    versionFile = os.path.join(path if path is not None else get_webots_home(), 
+                               'resources', 'version.txt')
     if not os.path.isfile(versionFile):
         return None
     with open(versionFile, 'r') as f:

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -51,7 +51,7 @@ def get_webots_home(version=get_required_webots_version()):
     environVariables = ['ROS2_WEBOTS_HOME', 'WEBOTS_HOME']
     for variable in environVariables:
         if variable in os.environ and os.path.isdir(os.environ[variable]) and \
-            get_webots_version(os.environ[variable]) == version:
+                get_webots_version(os.environ[variable]) == version:
             os.environ['WEBOTS_HOME'] = os.environ[variable]
             return os.environ[variable]
     # then using the 'which' command

--- a/webots_ros2_core/webots_ros2_core/utils.py
+++ b/webots_ros2_core/webots_ros2_core/utils.py
@@ -45,11 +45,9 @@ def get_required_webots_version_short():
     return make_short_version(get_required_webots_version())
 
 
-def get_webots_home(version=None):
+def get_webots_home(version=get_required_webots_version()):
     """Path to the Webots installation directory."""
     # search first with the environment variables
-    if version is None:
-        version = get_required_webots_version()
     environVariables = ['ROS2_WEBOTS_HOME', 'WEBOTS_HOME']
     for variable in environVariables:
         if variable in os.environ and os.path.isdir(variable) and get_webots_version(os.environ[variable]) == version:

